### PR TITLE
[qa] Added a check for empty line at the end of a file #29

### LINKED ISSUE #29: [qa] Add a check for empty line at the end of a file
We should add another automated test to check if all the files have an empty line at the end of the file.
Many GCI students are making this mistake and an automated test will help them in my opinion! :smile: 

PS: **binary files must be ignored**, the check should be performed only on text files.
If it can be performed with just a series of shell commands, we can use that and start adding this check to the `.travis.yml` build of this repo

### PR DESCRIPTION
**Pull request by ppabcd**
Fixes #29

### CODE REVIEW

**Comment by unsuitable001** [comment]
Good job @ppabcd. I'd recommend you to exclude "*.egg-info" instead of "openwisp-utils.egg-info"

**Comment by ppabcd** [comment]
i think it's will make errors. @unsuitable001

**Comment by unsuitable001** [comment]
@ppabcd I don't think so. I did the task by interacting between shell and python. And, I used the *egg-info. It never raised any error.
But, it's better to check yourself :). You can also check mine (un-merged) changes.

**Comment by ppabcd** [comment]
gambar
i get this error when using "*.egg-info"

**Comment by unsuitable001** [comment]
Put a / in front

**Comment by ppabcd** [comment]
get same error @unsuitable001

**Comment by unsuitable001** [comment]
@ppabcd a final try. `*/*egg-info/*`
If even this doesn't work, we'll talk after gci

**Comment by ppabcd** [comment]
i think it's works. Thank you @unsuitable001 .
And sorry if i resolve your pull.

**Comment by unsuitable001** [comment]
@ppabcd I'm glad to hear that it works. And, why sorry? We are all here to contribute :). Forget about competition. Working together, we'll make the best for us. :)

**Comment by ppabcd** [comment]
Your welcome, thank you too for @dwang and @unsuitable001 for help me improve this script.

**Comment by atb00ker** [comment]
Current output in case multiple files have error:
```
No new line A No new line B No new line C
```
I think this would be better.
```
No EOF line found in: A
No EOF line found in: B
No EOF line found in: C
```
or maybe:
```
No EOF line found in: 
A
B
C
```
Just a bit user friendly, what do you think? :smile:

**Comment by ppabcd** [comment]
i'll try with new line @atb00ker thank you.

**Comment by ppabcd** [comment]
Ok i'll waiting and add this changes to other repo when all done.

File: runcheckendline [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+set -e
+
+no_new_line=$(find . -type f -not -path '*/\.*' \
+    -not -path "*/*egg-info/*" \
+    -exec grep -Iq . {} \; -and -print0 \
+    -not -empty | xargs -0 -L1 bash -c \
+    'test "$(tail -c 1 "$0")" && echo "$0\n" || true')
+trim_new_line=$(echo "$no_new_line" | cut -f1 -d-)
+if [[ $trim_new_line == "" ]]; then
+    echo "All check done"
```
Comment (nemesisdesign) [inline runcheckendline]: can you remove this line please? It's not needed, see [Art of Unix programming, rule of silence](http://www.catb.org/esr/writings/taoup/html/ch01s06.html#id2878450).

Comment (nemesisdesign) [reply runcheckendline]: PS: can you also rebase this branch on the current master please?

Comment (ppabcd) [reply runcheckendline]: which line i must remove? 
8/9/10/11?

Comment (dwang) [reply runcheckendline]: I think @nemesisdesign wants you to remove the echo at the end (line 11), referring to the website linked above. "When a program has nothing surprising to say, it should say nothing."

Comment (nemesisdesign) [reply runcheckendline]: correct, I'm referring to `echo "All check done"`, an inline comment on github almost always refers to the line above the comment, unless otherwise stated

Comment (ppabcd) [reply runcheckendline]: i've done fix this line and update my master branch.

Comment (dwang) [reply runcheckendline]: I think you still need to rebase your branch with `master` in `openwisp-utils`. If you look at your repo right now, you will see this:
image
After you are done rebasing, it should look like:
image
If you have any questions, I'm happy to help 😃

Comment (nemesisdesign) [reply runcheckendline]: @ppabcd @dwang is right, you still need to update this branch with the master branch, you should be able to do that with:

```
git rebase origin/master
```

**Comment by ppabcd** [comment]
all done @nemesisdesign

**Comment by nemesisdesign** [comment]
We could try adding this check in other modules too, what do you think @atb00ker? Low priority though.

**Comment by nemesisdesign** [comment]
@ppabcd @atb00ker what do you think about extracting this new line check to something like a gist and having travis download and execute it? It may be easy to edit in case we need to do so. Otherwise if we add it to many modules and we find a flaw, we'll have to update it in all the modules.

**Comment by unsuitable001** [comment]
Should I close my alternative PR at https://github.com/openwisp/openwisp-utils/pull/38 as It is closed now? Or, leave it as it is?

**Comment by ppabcd** [comment]
hmm like edit one and you can change every repo? @nemesisdesign

**Comment by nemesisdesign** [comment]
yes @ppabcd 
yes I think so @unsuitable001, unless you have a specific improvement you want to propose, in that case you can update your PR

**Comment by ppabcd** [comment]
i think it's great idea without added more pull request on another repository @nemesisdesign

**Comment by unsuitable001** [comment]
You can say it as a simplified version @nemesisdesign  . Nothing more than that. I can remove some `echo` lines. This will make the script even shorter... so, what should I do? I need suggestion.

**Comment by nemesisdesign** [comment]
@unsuitable001 go for it

**Comment by unsuitable001** [comment]
It's unclear to me... 'go' to close? or, 'go' to remove the extra lines?

**Comment by nemesisdesign** [comment]
@unsuitable001 remove the extra lines

**Comment by unsuitable001** [comment]
@nemesisdesign Done. check it at #38

**Comment by atb00ker** [comment]
> @ppabcd @atb00ker what do you think about extracting this new line check to something like a gist and having travis download and execute it? It may be easy to edit in case we need to do so. Otherwise if we add it to many modules and we find a flaw, we'll have to update it in all the modules.

I tried using Gist for some other repositories and it doesn't work very well! Only the owner of the Gist can make changes in Gists.
If we need to download the latest version for travis, i think we can do that directly from the repository, example:
Consider [this](https://raw.githubusercontent.com/openwisp/openwisp-utils/master/runcheckendline
) link for `empty_line_check` file. It will always point to the latest version and if we want to point to a specfic version, we can use something like [this](https://raw.githubusercontent.com/openwisp/openwisp-utils/eb1e0e05546e01153cc2199b1aaaf29fcfc888e8/runcheckendline)** link.

However, we have done this cycle of making a change here and updating the repositories for 3 times in past 2-3 months and it's consuming too much time. I was thinking of the following to solve the problem:
Make a common file that will call all these seperate check files. We should call that file in travis of the other repositories so that from the next time when we make a change in that file it will reflect automatically in all the repositories. We can have the download link to latest version of that file in travis so that we don't need to update anything for ci.

What do you guys think about it?

** press `y` on GitHub to get the permanent link to a file.

**Comment by nemesisdesign** [comment]
@atb00ker you're right about gist, we could make a script file here in `openwisp-utils` that would allow more people to maintain it
